ICU-21117 Use Bazel to automate generation of Unicode data files

echeran opened the pull request:
##### Checklist

- [X] Issue filed: https://unicode-org.atlassian.net/browse/ICU-21117
- [X] Updated PR title and link in previous line to include Issue number
- [X] Issue accepted
- [ ] Tests included
- [ ] Documentation is changed or added

markusicu commented on icu4c/source/common/BUILD:
```diff
@@ -92,6 +92,170 @@ cc_library(
+)
+
+cc_library(
+    name = "uniset",
```
In toolutil, I expect that most cc_library targets want to have a single .cpp file each.

In common and i18n, this approach should also work, but may not be necessary. My hunch is that making one cc_library target for each dependencies.txt group should be sufficiently granular.

E.g.
```
group: uniset_core
    unifilt.o unifunct.o
    uniset.o bmpset.o unisetspan.o
  deps
    patternprops
    icu_utility
    uvector
```

echeran replied:
Done.

echeran commented:
@markusicu I updated the file copyright and description headers, and I added scripts to create and delete generated files. Can you verify that `common/norm2_nfc_data.h` is still being created correctly, and tell me if the scripts behave as intended?

markusicu commented on WORKSPACE:
```diff
@@ -0,0 +1,6 @@
+# © 2021 and later: Unicode, Inc. and others.
+# License & terms of use: http://www.unicode.org/copyright.html
+
+# A Bazel project is associated with directory the files and subdirectories
+# contained within. Each Bazel project requires a WORKSPACE file, which must
+# exist in the root directory of the project.
```
Trouble parsing this... How about
```suggestion
# This file tells Bazel where the root directory of the project is.
# "Target" and file paths in Bazel BUILD files are relative to this root directory.
```

markusicu commented on icu4c/source/common/BUILD:
```diff
@@ -0,0 +1,361 @@
+# © 2021 and later: Unicode, Inc. and others.
+# License & terms of use: http://www.unicode.org/copyright.html
+
+# This file defines Bazel targets for ICU4C header and source files that are
+# used across i18n algorithm implementations and related tools.
+# The configuration of dependencies among targets is strongly assisted by the
+# file in deptest that maintains such information, at
+# icu4c/source/test/depstest/dependencies.txt .
```
```suggestion
# This file defines Bazel targets for a subset of ICU4C "common" library header and source files.
# The configuration of dependencies among targets is strongly assisted by the
# file in depstest that maintains such information, at
# icu4c/source/test/depstest/dependencies.txt .
```

markusicu commented on icu4c/source/common/BUILD:
```diff
@@ -0,0 +1,361 @@
+        "unicode/*.h", # public
+        "*.h",         # internal
+        ],
+        exclude = ["norm2_nfc_data.h"],
```
Please add a comment like
```
# Instead of this checked-in file, our Bazel build process regenerates this file and then uses the new version.
```

echeran replied:
Done.

markusicu commented on icu4c/source/common/BUILD:
```diff
@@ -0,0 +1,361 @@
+    ],
+    local_defines = [
+        "U_COMMON_IMPLEMENTATION",
+        "U_ATTRIBUTE_DEPRECATED=",
```
Please add a comment like
```
# Linux platform defines.
```

markusicu replied:
Why do we need to define U_ATTRIBUTE_DEPRECATED as empty? umachine.h defines it for gcc and MSVC and otherwise leaves it empty anyway. And it is indirectly used in definitions across the libraries, so defining it only for "platform" is inconsistent.

markusicu commented on icu4c/source/common/BUILD:
```diff
@@ -0,0 +1,361 @@
+cc_library(
+    name = "uhash",
+    hdrs = [
+        # "hash.h",
```
please remove

markusicu commented on icu4c/source/common/BUILD:
```diff
@@ -0,0 +1,361 @@
+    includes = ["."],
+    hdrs = [
+        "normalizer2impl.h",
+        "norm2_nfc_data.h",
```
```suggestion
        "norm2_nfc_data.h",  # generated by gennorm2
```

markusicu commented on icu4c/source/common/BUILD:
```diff
@@ -0,0 +1,361 @@
+
+
+
+# Note: Bazel has chosen specifically to disallow any official support to break
+# through these constraints that would copying to or generating to source
+# directories: https://stackoverflow.com/questions/51283183/how-to-generate-files-in-source-folder-using-bazel
```
```suggestion
# Note: Bazel has chosen specifically to disallow any official support for breaking
# through constraints that would allow copying to or generating to source
# directories: https://stackoverflow.com/questions/51283183/how-to-generate-files-in-source-folder-using-bazel
```

markusicu commented on icu4c/source/data/unidata/delete_build_headers.sh:
```diff
@@ -0,0 +1,20 @@
+# © 2021 and later: Unicode, Inc. and others.
+# License & terms of use: http://www.unicode.org/copyright.html
+
+# This script deletes newly created header files produced by create_build_headers.sh ,
```
This is not quite what I imagined.

First, I assumed that we would be accumulating a list of generated files. "bazelisk clean" is much better!

Second, I suggested deleting the to-be-generated files before "bazelisk build" to ensure that the build process cannot rely on the to-be-generated files while generating. Therefore, we don't want to "git restore" them either. "git restore" seems only useful when something went wrong and we want to get back to a source tree that works with the "normal" build system.

I suggest you start a TODO section near the top of unidata/changes.txt and put the "bazelisk clean" line there with a one-liner about deleting the to-be-generated files. We can clean that up later.

As for "git restore", can you just `git restore .` (or something) at top level and ? If so, we don't need this .sh file here at all.

Or do we need a script with a hardcoded list? In that case, I would rename this .sh to restore_output_files.sh or simply restore.sh.

markusicu replied:
Reviewed together; I had still misunderstood. "bazelisk clean" deletes the *output* files, not the files in the source tree. We want to delete the latter, to guarantee that the bazel build does not pick up old versions of files that it generates.

Agreed on
- rename to clean.sh
- no "bazelisk clean" here but might mention it in unidata/changes.txt
- change "git restore" to "rm"

markusicu commented on icu4c/source/data/unidata/create_build_headers.sh:
```diff
@@ -0,0 +1,17 @@
+# © 2021 and later: Unicode, Inc. and others.
+# License & terms of use: http://www.unicode.org/copyright.html
+
+# This script runs Bazel to create (generate) header files from data that are
```
Not just header files. I suggest renaming this .sh to generate_output_files.sh or simply generate.sh.

markusicu commented on icu4c/source/data/unidata/norm2/BUILD:
```diff
@@ -0,0 +1,14 @@
+    default_visibility = ["//visibility:public"],
+)
+
+# automatically export plain files as targets in the package
```
plain?
plain text?

markusicu commented on tools/unicode/c/genprops/BUILD:
```diff
@@ -0,0 +1,50 @@
+        "//icu4c/source/tools/toolutil:uparse",
+        "//icu4c/source/tools/toolutil:toolutil",
+        "//icu4c/source/tools/toolutil:denseranges",
+        
```
In several targets you have empty lines just before the closing bracket. Why?

markusicu commented on icu4c/source/data/unidata/create_build_headers.sh:
```diff
@@ -0,0 +1,17 @@
+bazelisk build //icu4c/source/common:normalizer2
+cp $ICU_SRC/bazel-bin/icu4c/source/common/norm2_nfc_data.h $ICU_SRC/icu4c/source/common/norm2_nfc_data.h
+
+bazelisk run //tools/unicode/c/genprops $ICU_SRC/icu4c/
```
Please add commands to copy the outputs from genprops back into $ICU_SRC.

markusicu replied:
Actually, for now, genprops takes one folder for both input and output. It will write into the source tree, despite what Bazel likes. Might work for now, at least until we start also building and invoking genuca. Eventually genprops might need another parameter for the output folder.

markusicu commented on icu4c/source/common/BUILD:
```diff
@@ -0,0 +1,361 @@
+    local_defines = [
+        "U_COMMON_IMPLEMENTATION",
+        "U_ATTRIBUTE_DEPRECATED=",
+        "U_HAVE_XLOCALE_H=0",
+        "U_HAVE_STRTOD_L=1",
+        "U_HAVE_ELF_H=1",
+        "_REENTRANT",
```
I checked on these four. They are either not used anywhere, or not used in the common library, or documented as being obsolete. Let's remove them.

Details: https://unicode-org.atlassian.net/browse/ICU-21553

echeran reviewed (commented):
I believe I have applied all the changes as indicated in the PR and/or as discussed together. The only divergence is a minor one -- I put the comment `# Linux platform headers` in the first target, where other similar comments exist, not in the second target.

markusicu commented on icu4c/source/common/BUILD:
```diff
@@ -27,13 +26,16 @@ cc_library(
+    # Bazel workspace:
     # https://stackoverflow.com/a/65635893/2077918
     includes = ["."],
+    # Linux platform defines.
```
... except the one that's left here is not a Linux platform define. Please remove.

markusicu commented on icu4c/source/data/unidata/changes.txt:
```diff
@@ -49,6 +49,12 @@ For new script codes see http://www.unicode.org/iso15924/codechanges.html
 
 ---------------------------------------------------------------------------- ***
 
+Unicode 14.0 update for ICU 69
```
```suggestion
Unicode 14.0 update for ICU 70
```

markusicu commented on icu4c/source/data/unidata/clean.sh:
```diff
@@ -0,0 +1,17 @@
+# © 2021 and later: Unicode, Inc. and others.
+# License & terms of use: http://www.unicode.org/copyright.html
+
+# This script deletes the header files that generate.sh tries to regenerate.
```
```suggestion
# This script deletes files that generate.sh tries to regenerate.
```

markusicu commented:
... and once you have made those three fixes feel free to squash right away.
How about something like this for a commit comment? --> "ICU-21117 Use Bazel to automate generation of Unicode data files"

echeran commented:
Those changes sounds good, and I applied them with the suggested commit message for the squash+push.

markusicu reviewed (approved):
lgtm tnx
